source/sessions/driver-sessions.rst

```diff
@@ -1197,6 +1212,25 @@ Instead, we require users to pass session as a parameter to each function::
   collection.insertOne(session, {})
   session.endSession()
 
+Why does a network error cause the ``ServerSession`` to be discarded from the pool?
+-----------------------------------------------------------------------------------
+
+When a network error is encountered when executing an operation with a
+``ClientSession``, the operation may be left running on the server. Re-using
+this ``ServerSession`` can lead to parallel operations which violates the
+rule that a session must be used sequentially. This results in multiple
+problems:
+
+#. killSessions to end an earlier operation would surprisingly also end a
+   later operation.
+#. An otherwise unrelated operation that just happens to use that same server
+   session will potentially block waiting for the previous operation to
+   complete. For example, a transactional write will block a subsequent
+   transactional write.
+
+TODO: Justify why we allow applications to continue using an explicit session
+after a network error.
```

Comment:
I understand why we might want to never re-use a dirty session for implicit use, but I don't see the problem with allowing applications to continue using it for explicit use. Quoting your previous comment from the OP:

> This approach still allows parallel operations to be running on the server when using explicit sessions which isn't great but at least they will be related operations. Some (in my opinion bad) alternatives:
> * Drivers can raise an error when a dirty session is used for a later operation. I don't like this idea because how would an application recover from this kind of error?
> * A ClientSession can discard its dirty ServerSession and replace it with a new ServerSession from the pool. This seems wonky and I'm not sure it work be correct with respect to transactions.

Based on your earlier text, I might assume that the server prohibits concurrent execution of multiple operations using the same session (regardless of transactions).

I don't think we need to prohibit future use of dirty sessions, and I agree it would be disruptive for an application to deal with that. If the previous operation utilized retryable read/write behavior after a socket error and ultimately succeeded, the app shouldn't have to worry about anything blocking. If the previous operation failed completely, the app would have observed a socket error and is in a place to infer that future operations with the same session might block.

Silently replacing the ClientSession's underlying ServerSession does not seem kosher, as it violates the previous expectation that operations with the same ClientSession will all use the same LSID.

Reply:
> but I don't see the problem with allowing applications to continue using it for explicit use.

I am also advocating that we allow applications to continue using explicit dirty sessions. I just wanted to highlight that this can still lead to parallel operations on the server and get everyone on the same page. Sorry if that wasn't clear.

> I don't think we need to prohibit future use of dirty sessions, and I agree it would be disruptive for an application to deal with that.
> Silently replacing the ClientSession's underlying ServerSession does not seem kosher, as it violates the previous expectation that operations with the same ClientSession will all use the same LSID.

Agreed on both points.